hypershift-operator/controllers/manifests/controlplaneoperator/manifests.go

```diff
@@ -71,11 +71,11 @@ func CAPICluster(controlPlaneOperatorNamespace string, infraID string) *capiv1.C
 	}
 }
 
-func HostedControlPlane(controlPlaneNamespace string, hostedClusterName string) *hyperv1.HostedControlPlane {
+func HostedControlPlane(namespace string, name string) *hyperv1.HostedControlPlane {
```

Comment:
What motivates this change? The argument names were intentionally clear before that the HCP instance is derived from the HostedCluster name — now that's not clear and it also looks like the name has been changed to be a function of the HostedCluster's infra ID and I'm not sure why that's desired or necessary

Reply:
hostedClusterName string seemed like a false contract to me. As a consumer of the function I can pass any string with this signature regardless of the variable naming. We just happen to choose to pass the hostedClusterName so far though there's no programatic requirement for it. Now is convenient to pass the infraID as the string because of https://github.com/openshift/hypershift/pull/246#discussion_r633867308. Let's keep the discussion in thread below to avoid having it scattered.

Reply:
I addressed all comments and we are back to passing the `hostedCluster` name as the input when calling this func, however I'm still in favour of this particular change because of reasons in my comment the above. If we think the name for the `HostedControlPlane` is required to be the `hostedCluster` name, then we must enforce that programatically and pass the `hostedCluster` struct as parameter in the signature function for `HostedControlPlane()`, otherwise this is just a string, the func signature name does not define nor enforce any particular input other than just the `name` for a `HostedControlPlane`.

Reply:
Apologies in advance for the long reply, as I started thinking about it the issues are more subtle than I expect and I was reminded of our past experiences.

>hostedClusterName string seemed like a false contract to me. As a consumer of the function I can pass any string with this signature regardless of the variable naming. We just happen to choose to pass the hostedClusterName so far though there's no programatic requirement for it. 

Go doesn't leave much choice here without a lot of silly work — under the current design of these functions, the input here is actually more like "HostedClusterMetadata" but I have no good way of representing that type without generics. For context, these manifest functions are intentionally designed to be able to reconstruct the coordinates for all the resources in the tree with given _only the metadata_ for each node. Otherwise the function seems to me to become more like a reconciliation function.

>If we think the name for the HostedControlPlane is required to be the hostedCluster name, then we must enforce that programatically and pass the hostedCluster struct as parameter in the signature function for HostedControlPlane()

It's an interesting point. We actually did use this style at one time, although never leaked non-metadata into the calculations for naming (re: the design intent above). 

One issue with the approach of working with non-metadata details (i.e. accepting the whole HC) is that it means you need a fully valid HC (or other resource for the other functions) to determine the _coordinates_ of descendent resources, even though the use case doesn't even assume these resources still exist. If you no longer have the HC itself (but still have its metadata, e.g. namespace/name) you can find the coordinates to its related resources.

Consider: If I have the metadata for a HostedCluster, I can look up the HostedControlPlane's deployment through the HostedControlPlane even if the HostedControlPlane was deleted somehow, because for that chain the coordinates are ultimately a function of the HostedCluster's metadata. If I needed the full HCP itself (e.g. a field on spec) to find the coordinates for the HostedControlPlane deployment, I've lost the ability to address that node (the deployment) without reconstructing the HostedControlPlane resource. 

One last thing is that accepting the whole object as input in these functions does make sense on the one hand as an abstraction point to insulate from changing how we address things, but on the other hand it's a weak contract _unless you clearly require the full object_ (otherwise you have to specify in docs or something which subset of fields to populate, which is no better than what's already here re: named arguments). I think that goes back to the lack of typed metadata per object point.

P.S., this change made me consider that at some point (post release?) we'll have to carefully consider any changes to any of the manifest stuff because of the migration implications; if we change a resource's name, unless we migrate existing resources in advance, bad things will probably happen (and we also have to clean up the "old" names).